# DTSC 580: Data Manipulation
## Assignment: LEGO Project
### Name:

![LegoSchema](Lego_Schema.PNG)

## Overview

In this assignment, you will load the various tables from the [Rebrickable](https://rebrickable.com/) website and perform analysis on the data.  Rebrickable is a fantastic website that is used to show you which LEGO sets you can build from the sets and parts you already own.  They will even show you which parts you are missing to complete the new set and suggest sets that you could purchase to get those parts.  

## Data

The data has been downloaded directly from their website and includes files for the tables shown in the schema above.  Your first task will be to go through the data and review the schema to understand how the tables all fit together.  There is no data dictionary that I am aware of, but you should be able to understand this data by reviewing the files and schema and visiting their [Download](https://rebrickable.com/downloads/) page and [API](https://rebrickable.com/api/) documentation.  **Note that you must use the files downloaded from Brightspace for this assignment and not download new files from the Rebrickable website.**

## Assignment

In this assignment, you will create a number of functions to complete certain tasks.  This is not needed if you were working on this project on your own, but since we want to see if you can select the data via Pandas/Numpy, we will use those functions in the automatic grading in CodeGrade.  Without this, students could try to look up the answers in the `csv` files instead of wrangling the data with Python/Pandas, which would defeat the purpose of this class.

In addition, we would strongly suggest that you merge data sets as appropriate for the given task.  For example, you might merge the `sets` and `themes` DataFrames to answer one question, and then merge `sets`, `inventories`, `inventory_minifigs`, and `minifigs` to answer another.  This will make your code much easier to write and will be key in working with data across multiple tables.  In fact, I was able to create each function with only one line of code because of the merging of certain data sets.  Also note that you may not use all the tables in this assignment.  This is also an important skill for a Data Scientist to know when to use data and what data is unnecessary.

When complete, save your notebook as `lego.ipynb` and submit to CodeGrade for automatic grading.

## Note

<u>Show Work</u>

Remember that you must show your work.  Students submissions are spot checked manually throughout the term to verify that they are not hard coding the answer from looking only in the file or in CodeGrade's expected output.  If this is seen, the student's answer will be manually marked wrong and their grade will be changed to reflect this.

For example, if the answer to Q1, the mean of a specific column, is 22:
```
# correct way
Q1 = df['column_name'].mean()

# incorrect way
Q1 = 22
```

Let's start out by importing some standard imports.

In [79]:
# common imports
import pandas as pd
import numpy as np

# Do not change this option; This allows the CodeGrade auto grading to function correctly
pd.set_option('display.max_columns', None)
import warnings
warnings.filterwarnings("ignore")

## Import Data

Load the various Lego datasets below calling them the same as the name of the table as seen in the schema.  Note that you may or may not have to use all the data sets.  One of the things that all Data Scientists and Analysts have to decide when working on a project is what data is useful for a task and what data is not.  Remember to look at the schema above to see how all the data sets are connected.

The autograder in CodeGrade will have the `csv` files in the same location where the notebook is run.  Because of this, please make sure your local files are in the same location as your notebook.

In [80]:
# import Lego data sets here
# make sure that files are in the same folder as your notebook
inventories = pd.read_csv('inventories.csv')
inventory_sets = pd.read_csv('inventory_sets.csv')
sets = pd.read_csv('sets.csv')
themes = pd.read_csv('themes.csv')
inventory_minifigs = pd.read_csv('inventory_minifigs.csv')
minifigs = pd.read_csv('minifigs.csv')
inventory_parts = pd.read_csv('inventory_parts.csv')
part_categories = pd.read_csv('part_categories.csv')
parts = pd.read_csv('parts.csv')
colors = pd.read_csv('colors.csv')
part_relationships = pd.read_csv('part_relationships.csv')
elements = pd.read_csv('elements.csv')

## Exploration

Take time to look at all of the data sets using `.head()`, `.info()`, `.describe()`, etc to familiarize yourself with the data sets.  As you look at them, notice how the various tables connect together.

In [81]:
### ENTER CODE HERE ###
sets.tail()

set_num                             name  year  theme_id  num_parts
18571      XWING-1              Mini X-Wing Fighter  2019       158         60
18572      XWING-2                X-Wing Trench Run  2019       158         52
18573  YODACHRON-1  Yoda Chronicles Promotional Set  2013       158        413
18574   YTERRIER-1                Yorkshire Terrier  2018       598          0
18575     ZX8000-1             ZX 8000 LEGO Sneaker  2020       501          0

## Assignment

Answer all the questions below using the available data.  There are multiple ways to get some of the answers.  I would suggest that you attempt to merge some data sets as needed to assist with getting answers.  This will make your tasks a lot easier and your code will be cleaner.

Also, you will be creating functions to get the answers to some of these questions.  This will make it much easier for you to get answers to similar questions in the future and CodeGrade will be checking your functions using different inputs.  

**Q1:** Create a function called `data_count` that takes as input the data set (for example: `sets`) and returns the total number of rows in the data (as an integer).  Nothing else should be returned.

In [82]:
### ENTER CODE HERE ###
data_count= lambda x: x.shape[0]

**Code Check:**  If you call your `data_count` function using the `sets` DataFrame, you should see that there are 18,576 sets represented in the data.

In [83]:
### CODE CHECK ###
data_count(sets)

18576

**Q2:** What is the average number of parts for a Lego set (rounded to the nearest integer)?  Save this as `Q2`.

In [114]:
### ENTER CODE HERE ###
Q2= sets['num_parts'].mean()
Q2=round(Q2)
Q2

161

**Q3:** What is the median number of parts per set (output as an integer)?  Save this as `Q3`.

In [115]:
### ENTER CODE HERE ###
Q3= sets['num_parts'].median()
Q3=round(Q3)
Q3

37

You should see that the median number of parts is much lower than the average number of parts per set.  This tells us that there are some Lego sets with a very high number of parts that brings the average up.

**Q4:** What is the largest number of parts that a set has? Save this as `Q4`.

In [86]:
### ENTER CODE HERE ###
Q4= sets['num_parts'].max()
Q4

11695

**Q5:** Given the set with the largest number of parts calculated above, select its row from the `sets` DataFrame.  Save this as `Q5`.  Do not reset the index.

In [87]:
### ENTER CODE HERE ###
Q5=sets[sets['num_parts']==Q4]
Q5

set_num       name  year  theme_id  num_parts
3867  31203-1  World Map  2021       709      11695

**Q6:** Create a function called `select_set_row` where you pass it the `set_num` (as a string) and the function returns the respective row of the DataFrame from `sets`. Do not reset the index.

In [88]:
### ENTER CODE HERE ###
def select_set_row(set_num):
    return sets[sets['set_num']==set_num]

**Q7:** Create a function called `select_set_numparts` that takes as input a specific number of parts (as an integer) and selects all the rows in the `sets` DataFrame that contain those number of parts.  Return the DataFrame sorted by `year` and then by `set_num`.  Do not reset the index.

In [89]:
### ENTER CODE HERE ###
def select_set_numparts(num_parts):
    return sets[sets['num_parts']==num_parts].sort_values(by=['year','set_num'])

**Code Check:** Call the `select_set_numparts()` function to select all sets with `0` number of parts.  There should be 3,146 rows returned.

In [90]:
### CODE CHECK ###
select_set_numparts(0)

set_num                                               name  \
11061          700.1-2                     Extra-Large Gift Set (Mursten)   
17563         b55dk-01               Lego Mursten - System i Leg Byggebog   
17564         b56de-01                                    System im Spiel   
11072          700.2-2                                       Gift Package   
11078         700.3A-2                                       Gift Package   
...                ...                                                ...   
15619         854195-1                                    Ernie Key Chain   
15620         854197-1                               Demogorgon Key Chain   
17170  9780744054583-1  DC Comics Super Heroes: Character Encyclopedia...   
17193  9780794447557-1                               City: Color The City   
17200  9780794448172-1           DC Comics Super Heroes: Ready For Action   

       year  theme_id  num_parts  
11061  1953       366          0  
17563  1955       497          0  
17564  1956       497          0  
11072  1957       366          0  
11078  1957       372          0  
...     ...       ...        ...  
15619  2022       503          0  
15620  2022       503          0  
17170  2022       497          0  
17193  2022       497          0  
17200  2022       497          0  

[3146 rows x 5 columns]

**Q8:**
- The oldest set listed was made in what year?  Save the year as `Q8A` (as an integer).  
- The newest set listed was made in what year? Save the year as `Q8B` (as an integer).

In [91]:
### ENTER CODE HERE ###
Q8A= sets['year'].min()
Q8A

1949

In [92]:
### ENTER CODE HERE ###
Q8B= sets['year'].max()
Q8B

2022

**Q9:**
- Create a function called `select_set_year` that takes as input a year (as an integer) and returns the rows from the `sets` DataFrame that matches that year.  Sort the DataFrame by `set_num`.  Do not reset the index.

In [93]:
### ENTER CODE HERE ###
def select_set_year(year):
    return sets[sets['year']==year].sort_values(by='set_num')

**Q10:**
- Create a function called `theme_by_year` that takes as input a year (as an integer) and shows the theme ids and theme names (listed in order by theme id) that were in sets that year.  
- The column names must be `id` and `name_themes` (to differentiate between the name of a theme and the name of a set) in that order.  
- The index should be reset and go from `0` to `n-1`.  
- Each theme should only be listed once even if it appeared in more than one set from that year -- duplicate themes should be based on theme id and not name since there are some themes with the same name but with a different id.
- Hint:  It will help if you were to think about merging appropriate DataFrames to help you get this answer.

In [117]:
### ENTER CODE HERE ###
def theme_by_year(year):
    merged_df = sets.merge(themes, left_on='theme_id', right_on='id',suffixes=('_sets', '_themes') )
    return merged_df[merged_df['year']==year][['theme_id','name_themes']].drop_duplicates().sort_values(by='theme_id').reset_index(drop=True)


**Code Check:** Call the `theme_by_year()` function on the year `1960`.  Your output should look as follows:

| | id | name_themes|
|-|---|---|
|0|371|Supplemental|
|1|497|Books|
|2|513|Classic|

**Optional Practice:** Call the `theme_by_year()` function using your birth year to see the Lego themes that were popular when you were born.

In [119]:
### CODE CHECK / OPTIONAL PRACTICE ###
theme_by_year(1978)

theme_id               name_themes
0          4            Expert Builder
1         67              Classic Town
2        189            Classic Castle
3        227                 Christmas
4        235                      4.5V
5        363                      Boat
6        364  Building Set with People
7        404                Hobby Sets
8        405                 Homemaker
9        412                   Airport
10       453                   Technic
11       469                 Basic Set
12       473              Supplemental
13       501                      Gear
14       505                 Basic Set
15       632                      Town

**Q11:**
- Create a function called `theme_by_name` that takes as input a theme name and outputs all the sets (in `set_num` order) that had that theme.  
- The DataFrame should include the following columns in this order and with these exact names: `set_num`,`name_sets`,`year`,`num_parts`, `theme_id`, and `name_themes`.  
- The index should go from `0` to `n-1`.

In [96]:
### ENTER CODE HERE ###
def theme_by_name(theme_name):
    merged_df = sets.merge(themes, left_on='theme_id', right_on='id',suffixes=('_sets', '_themes') )
    return merged_df[merged_df['name_themes']==theme_name][['set_num','name_sets','year','num_parts', 'theme_id', 'name_themes']].sort_values(by='set_num').reset_index(drop=True)

**Code Check:** Call the `theme_by_name()` function using the theme name `Gear`.  You should return 1,904 sets that match that theme name.

In [97]:
### CODE CHECK ###
theme_by_name('Gear')

set_num                                          name_sets  \
0              11138-1                            Bionicle Barraki Wallet   
1       119645111436-1                Classic Space Astronaut Plush (Red)   
2       119645111504-1               Classic Space Astronaut Plush (Blue)   
3       119645111740-1              Classic Space Astronaut Plush (White)   
4       119645111818-1               Classic Space Astronaut Plush (Pink)   
...                ...                                                ...   
1899           WATCH-3                   Watch System Voyager (UFO Space)   
1900         WIIGAME-1                              Bionicle Heroes - Wii   
1901         WIIGAME-2  Indiana Jones: The Original Adventures Video G...   
1902  WOODENELEPHANT-1  Original Wooden Elephant Pull-Along Toy and Print   
1903          ZX8000-1                               ZX 8000 LEGO Sneaker   

      year  num_parts  theme_id name_themes  
0     2008          0       501        Gear  
1     2021          0       501        Gear  
2     2021          0       501        Gear  
3     2021          0       501        Gear  
4     2021          0       501        Gear  
...    ...        ...       ...         ...  
1899  1997         29       501        Gear  
1900  2007          0       501        Gear  
1901  2008          0       501        Gear  
1902  2021          0       501        Gear  
1903  2020          0       501        Gear  

[1904 rows x 6 columns]

**Q12:** Create a function called `theme_by_setnum` that takes as input the set_num (as a string) and returns the respective theme name as a string.

In [98]:
### ENTER CODE HERE ###
def theme_by_setnum(set_num):
    merged_df = sets.merge(themes, left_on='theme_id', right_on='id',suffixes=('_sets', '_themes') )
    return merged_df[merged_df['set_num']==set_num]['name_themes'].values[0]

**Code Check:** Visit the [Sets page](https://rebrickable.com/sets/) from the Rebrickable website and find a set that you like.  Call your `theme_by_setnum()` function on that set and make sure the correct theme name is output.

In [99]:
### CODE CHECK ###
theme_by_setnum('11138-1')

'Gear'

**Millennium Falcon Prep Exercise:** Find all the sets named 'Millennium Falcon'. Of these sets, find the set with the largest number of parts.  You will use this information for later questions.

In [100]:
### ENTER CODE HERE ###
millennium_falcon_sets = sets[sets['name']=='Millennium Falcon']
Q5_prep = millennium_falcon_sets[millennium_falcon_sets['num_parts']==millennium_falcon_sets['num_parts'].max()]
Q5_prep

set_num               name  year  theme_id  num_parts
12884  75192-1  Millennium Falcon  2017       171       7541

**Q13:**
- How many minifigs are contained in the set from the `Millennium Falcon Prep Exercise`?  Save this as `Q13` (as an integer).  
- Note that there are sometimes multiple versions of a set as shown in the `inventories` table.  This will not matter for this question as there is only one version for the respective set but this should be noted for future questions.
- Remember to show your work or your answer will manually get marked as incorrect.

In [101]:
### ENTER CODE HERE ###
merged_minifigs = inventory_minifigs.merge(inventories, left_on='inventory_id', right_on='id')
Q13 = merged_minifigs[merged_minifigs['set_num']=='75192-1'].shape[0]
Q13

8

**Q14:**
- Create a function called `get_minifigs` that takes as input a set number (as a string) and a version number (as an integer) and outputs all minifigs from that set / version combination.  
- The DataFrame should be in `fig_num` order with the following columns/column names: `set_num`, `fig_num`, `fig_name`, `fig_num_parts`.  
- The index should go from 0 to n-1.

In [102]:
### ENTER CODE HERE ###
def get_minifigs(set_num, version):
    merged_minifigs = inventory_minifigs.merge(inventories, left_on='inventory_id', right_on='id')
    merged_minifigs = merged_minifigs.merge(minifigs, on='fig_num')
    return merged_minifigs[(merged_minifigs['set_num']==set_num) & (merged_minifigs['version']==version)].sort_values(by='fig_num')[['set_num', 'fig_num', 'name', 'num_parts']].rename(columns={'name': 'fig_name', 'num_parts': 'fig_num_parts'}).reset_index(drop=True)

**Q15:** Using the above function, output information about the minifigs from the set discussed in the `Millennium Falcon Prep Exercise`.  Save this output as `Q15`.

In [103]:
### ENTER CODE HERE ###
Q15= get_minifigs('75192-1',1)
Q15

set_num     fig_num                                           fig_name  \
0  75192-1  fig-001714                              Chewbacca, Dark Brown   
1  75192-1  fig-001810                             Finn, Black Undershirt   
2  75192-1  fig-002057                                 Rey, Dark Tan Robe   
3  75192-1  fig-002514    C-3PO, Pearl Gold, Colorful Wires, Printed Legs   
4  75192-1  fig-002544                                               BB-8   
5  75192-1  fig-002549  Princess Leia, Hoth Outfit, White Jacket, Prin...   
6  75192-1  fig-002550  Han Solo, Dark Blue Jacket, Dark Brown Legs, P...   
7  75192-1  fig-002551                               Han Solo, Old, Angry   

   fig_num_parts  
0              3  
1              4  
2              4  
3              3  
4              2  
5              4  
6              4  
7              4

**Code Check:** Make sure that your `get_minifigs()` function works with both version 1 and version 2 of set 10217-1, Harry Potter's Diagon Alley.  Both versions should have 11 minifigs in the DataFrame.

In [104]:
### CODE CHECK ###
get_minifigs('10217-1',2)

set_num     fig_num                                           fig_name  \
0   10217-1  fig-002229                   Skeleton (Standard) (3626b Head)   
1   10217-1  fig-006075                 Rubeus Hagrid, Dark Brown Overcoat   
2   10217-1  fig-006084  Lucius Malfoy, Black Robes and Cape, Death Eat...   
3   10217-1  fig-006130  Harry Potter, Open Dark Blue Jacket over White...   
4   10217-1  fig-006138                                    Fenrir Greyback   
5   10217-1  fig-006162  Ron Weasley, Red Tartan Sweater Vest, Reddish ...   
6   10217-1  fig-006163          Fred / George Weasley in Dark Orange Suit   
7   10217-1  fig-006164                                 Garrick Ollivander   
8   10217-1  fig-006165    Goblin, Black Jacket and Light Bluish Gray Legs   
9   10217-1  fig-006166        Goblin, Black Jacket and Reddish Brown Legs   
10  10217-1  fig-006167  Hermione Granger, Light Bluish Gray Sweater wi...   

    fig_num_parts  
0               6  
1               3  
2               5  
3               4  
4               4  
5               4  
6               4  
7               4  
8               3  
9               3  
10              4

**Q16:**
- Create a function called `minifigs_from_themes` that takes as input a theme id (as an integer) and outputs all the minifigs contained in sets with that theme.  
- The DataFrame should be in `fig_num` order with the following columns/column names: `fig_num`, `fig_name`.
- The index should go from 0 to n-1.  
- Exact duplicates should be removed so that only unique figs are included.

In [105]:
### ENTER CODE HERE ###
def minifigs_from_themes(theme_id):
    merged_minifigs = inventory_minifigs.merge(inventories, left_on='inventory_id', right_on='id')
    merged_minifigs = merged_minifigs.merge(minifigs, on='fig_num')
    merged_minifigs = merged_minifigs.merge(sets, on='set_num')
    return merged_minifigs[merged_minifigs['theme_id']==theme_id][['fig_num', 'name_y']].drop_duplicates().sort_values(by='fig_num').rename(columns={'name_y': 'fig_name'}).reset_index(drop=True)

**Code Check:** Call the `minifigs_from_themes()` function using the theme number of the set discussed in the `Millennium Falcon Prep Exercise`.  There should be 149 minifigs output for that respective theme.

In [106]:
### CODE CHECK ###
minifigs_from_themes(sets[sets['set_num']=='75192-1']['theme_id'].values[0])

fig_num           fig_name
0    fig-000003         Death Star
1    fig-000124       Ewok Village
2    fig-000124         Death Star
3    fig-000124  Millennium Falcon
4    fig-000144        Sandcrawler
..          ...                ...
161  fig-011942              AT-AT
162  fig-011944              AT-AT
163  fig-011945              AT-AT
164  fig-011946              AT-AT
165  fig-011957              AT-AT

[166 rows x 2 columns]

**Q17:**
- Create a function called `sets_from_minifig` that takes as input a minifig number (as a string) and returns a list of sets a minifig has appeared in.  
- Return a DataFrame (sorted by `set_num` and then by `version`) with the following columns: `set_num`, `version`,`name_sets`, `fig_num`, `fig_name`.  
- The index should go from 0 to n-1.

In [122]:
### ENTER CODE HERE ###
def sets_from_minifig(fig_num):
    merged_minifigs = inventory_minifigs.merge(inventories, left_on='inventory_id', right_on='id')
    merged_minifigs = merged_minifigs.merge(minifigs, on='fig_num')
    merged_minifigs = merged_minifigs.merge(sets, on='set_num')
    return merged_minifigs[merged_minifigs['fig_num']==fig_num][['set_num', 'version','name_y', 'fig_num', 'name_x']].sort_values(by=['set_num', 'version']).rename(columns={'name_x': 'fig_name', 'name_y': 'name_sets'}).reset_index(drop=True)

**Code Check:** Call your `sets_from_minifig()` function on one of the Lego Friends minifigs, Andrea, who has figure number: `fig-001039`.  Your function should return a DataFrame that contains the following sets:

- Andrea on the Beach
- Birthday Party
- City Park Cafe

In [123]:
### CODE CHECK ###
sets_from_minifig('fig-001039')

set_num  version            name_sets     fig_num  \
0  30100-1        1  Andrea on the Beach  fig-001039   
1  30107-1        1       Birthday Party  fig-001039   
2   3061-1        1       City Park Café  fig-001039   

                                            fig_name  
0  Andrea - Bright Light Orange Top, Light Aqua S...  
1  Andrea - Bright Light Orange Top, Light Aqua S...  
2  Andrea - Bright Light Orange Top, Light Aqua S...

**Q18:**

*Note:  Before you begin working on this question, please note that CodeGrade has a 1GB RAM limit.  If you see an `Exit -9` code when submitting your assignment, it is probably due to this RAM limit and previous students had an issue with this question causing too much memory usage.  The biggest thing to think about is to only use the tables that you need to answer for this question.  For instance, you do not need any of the minifigs data since this is not needed for the output and the merging of this data could cause the number of rows to "explode" in terms of the total count.  I understand that this can be annoying that we have to work around CodeGrade's limitations, but it is good practice in memory management.*

*You should only need to merge the following datasets for this answer so as to not go above the RAM limit: `themes`, `sets`, `inventories`, `inventory_parts`, `parts`, and `colors`*

- Create a function called `set_parts` that takes as input the set number (as a string) and version number (as an integer) and outputs a DataFrame with each part in the set.  
- The DataFrame should have the following columns/column names: `set_num`, `version`, `name_sets`, `year`, `name_themes`, `part_num`, `name_part`, `name_color`,`quantity`,`is_spare`.  
- The Data should be sorted by `part_num` and then `name_color`.  
- The index should go from 0 to n-1.

In [109]:
### ENTER CODE HERE ###
def set_parts(set_num, version):
    # Merge sets and inventories first to keep inventory_id
    merged_df = sets.merge(inventories, left_on='set_num', right_on='set_num')
    # Filter by set_num and version early to reduce DataFrame size
    merged_df = merged_df[(merged_df['set_num']==set_num) & (merged_df['version']==version)]
    # Merge with themes
    merged_df = merged_df.merge(themes, left_on='theme_id', right_on='id', suffixes=('_sets', '_themes') )
    # Merge with inventory_parts using the 'id' from inventories
    merged_df = merged_df.merge(inventory_parts, left_on='id_sets', right_on='inventory_id')
    # Merge with parts and rename the part name column
    merged_df = merged_df.merge(parts, left_on='part_num', right_on='part_num')
    merged_df = merged_df.rename(columns={'name': 'name_part'})
    # Merge with colors and rename the color name column
    merged_df = merged_df.merge(colors, left_on='color_id', right_on='id')
    merged_df = merged_df.rename(columns={'name': 'name_color'})


    # Select and rename columns as per the problem description
    result_df = merged_df[['set_num', 'version', 'name_sets', 'year', 'name_themes', 'part_num', 'name_part', 'name_color','quantity','is_spare']]

    # Sort the DataFrame
    result_df = result_df.sort_values(by=['part_num', 'name_color']).reset_index(drop=True)

    return result_df

**Code Check**: Using the `set_parts()` function, output a list of parts for the set discussed in the `Millennium Falcon Prep Exercise`. There should be a total of 730 rows in the DataFrame, which includes 676 parts and 54 spare parts. You can check the respective page on the rebrickable website to verify these results.

In [110]:
### CODE CHECK ###
set_parts('75192-1',1)

set_num  version          name_sets  year                name_themes  \
0    75192-1        1  Millennium Falcon  2017  Ultimate Collector Series   
1    75192-1        1  Millennium Falcon  2017  Ultimate Collector Series   
2    75192-1        1  Millennium Falcon  2017  Ultimate Collector Series   
3    75192-1        1  Millennium Falcon  2017  Ultimate Collector Series   
4    75192-1        1  Millennium Falcon  2017  Ultimate Collector Series   
..       ...      ...                ...   ...                        ...   
725  75192-1        1  Millennium Falcon  2017  Ultimate Collector Series   
726  75192-1        1  Millennium Falcon  2017  Ultimate Collector Series   
727  75192-1        1  Millennium Falcon  2017  Ultimate Collector Series   
728  75192-1        1  Millennium Falcon  2017  Ultimate Collector Series   
729  75192-1        1  Millennium Falcon  2017  Ultimate Collector Series   

    part_num                                          name_part  \
0      10197  Technic Pin Connector Hub with 2 Perpendicular...   
1      10201            Bracket 1 x 2 - 1 x 4 [Rounded Corners]   
2      10247  Plate Special 2 x 2 with 1 Pin Hole [Complete ...   
3      10247  Plate Special 2 x 2 with 1 Pin Hole [Complete ...   
4      10314  Brick Curved 1 x 4 x 1 1/3 No Studs, Curved To...   
..       ...                                                ...   
725    99563  Tile Special 1 x 2 with Sloped Walls AKA Money...   
726    99780                     Bracket 1 x 2 - 1 x 2 Inverted   
727    99780                     Bracket 1 x 2 - 1 x 2 Inverted   
728    99781                              Bracket 1 x 2 - 1 x 2   
729    99784  Bar 1 x 12 with 1 x 2 Plate End with Hollow St...   

            name_color  quantity is_spare  
0     Dark Bluish Gray         3        f  
1    Light Bluish Gray        12        f  
2     Dark Bluish Gray         5        f  
3                White        22        f  
4                  Tan         3        f  
..                 ...       ...      ...  
725  Light Bluish Gray        70        f  
726              Black         6        f  
727   Dark Bluish Gray        25        f  
728  Light Bluish Gray         2        f  
729  Light Bluish Gray         3        f  

[730 rows x 10 columns]

## Optional Exercises
These exercises are optional and are not graded by CodeGrade.  They can be used as extra optional practice.

- What is the median year for the sets data?  Given the oldest and newest year in the data, what does this median value tell you about whether the number of sets have increased or decreased over the years?
- What minifig has the most number of parts?
- What minifig shows up in the most number of sets?
- When was the first minifig included in a set?
- What set has the most number of minifigs
- What part is used the most for all sets?
- What color is used the most for all parts?
- Get a list of all unique color names, sort by color name, and output as a list.
- How many total transparent colors are in the colors DataFrame?
- Create a function called get_part_colors that takes as input a part number (as a string) and returns a DataFrame of all colors a part has appeared in.
- What part / color combination is used the most for all sets?
- What part / color combination is used the least for all sets?
- Explore some of your favorite themes and which colors are used the most for those themes.
- Plot a line graph of the average number of parts per set per year.
